### PR TITLE
Implement decomposition for aten.tensor_split.tensor_indices_or_sections

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107251

Summary: Before this change, the tensor_indices_or_sections variant of aten.tensor_split causes a `RuntimeError: The tensor has a non-zero number of elements` due to that operation needing to introspect data. Decomposing into one of the other two tensor_split variants fixes the problem.

Test Plan:
Enabled tensor_split tests in test/inductor/test_torchinductor_opinfo.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1200,6 +1200,25 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
     return torch.split(self, split_sizes, dim)
 
 
+@aten.tensor_split.tensor_indices_or_sections.py_impl(DispatchKey.CompositeImplicitAutograd)
```
Comment (masnesral) [inline torch/_decomp/decompositions.py:1203]: @eellison , I couldn't discern if there's a convention for where this goes in the file, so I stuck it next to a couple other "split" decomps. ¯\_(ツ)_/¯

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1200,6 +1200,25 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
...
+    split_dim = tensor_indices_or_sections.dim()
+    assert split_dim == 1 or split_dim == 0, (
+        f"tensor_split expected tensor_indices_or_sections to be a zero-dimensional or "
+        "one-dimensional tensor, but got a tensor with {split_dim} dims"
+    )
+    if split_dim == 0:
+        sections = tensor_indices_or_sections.item()
+        return self.tensor_split(sections, dim)
+    else:
+        indices = [i.item() for i in tensor_indices_or_sections]
```
Comment (masnesral) [inline torch/_decomp/decompositions.py:1218]: @eellison Is this list comprehension going to be inefficient?

Comment (eellison) [reply torch/_decomp/decompositions.py:1221]: It's true that calling .item() in a loop is an anti-pattern. However, for our case currently, .item() will either return a constant, or will graph break, so it doesnt affect us. This tensor is also guaranteed to be on cpu so the `.item()` perf is not quite as bad as on cuda (although it's still bad).

The alternative to .item() in a loop is `.tolist()`. However, `tolist()` isn't supported for tensor subclasses (aka fake tensor). Maybe that should be an op ?

soliciting other opinions, cc @bdhirsh @ezyang

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -1200,6 +1200,28 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
...
+@aten.tensor_split.tensor_indices_or_sections.py_impl(
+    DispatchKey.CompositeImplicitAutograd
+)
+def tensor_split_tensor_indices_or_sections_py_impl(
+    self: Tensor,
+    tensor_indices_or_sections: Tensor,
+    dim: int = 0,
+) -> List[Tensor]:
+    split_dim = tensor_indices_or_sections.dim()
+    assert split_dim == 1 or split_dim == 0, (
```
Comment (eellison) [inline torch/_decomp/decompositions.py:1212]: nit: f strings will format eagerly, even if the assertion passes. we use `torch._check` to lazily format exception messages

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -1200,6 +1200,28 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> Tuple[Tensor, ...]:
...
+        "tensor_split expected tensor_indices_or_sections to be a zero-dimensional or "
+        f"one-dimensional tensor, but got a tensor with {split_dim} dims"
+    )
```
Comment (eellison) [inline torch/_decomp/decompositions.py:1215]: nit: check split_dtype is cpu and is kLong ?

**Review by eellison — COMMENTED** [review]
sweet, looks good ! I think the .item() calls are fine currently given, but lets wait for one other opinion.

**Review by ezyang — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**